### ISSUE TITLE
Model.fit displays wrong total count in progress bar when using validation data with undefined length

### ISSUE THREAD

**Issue by lojzezust** [tensorflow/tensorflow#34457]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04.6 LTS
- TensorFlow installed from (source or binary): binary (pip install tensorflow)
- TensorFlow version (use command below): v2.0.0-rc2-26-g64c3d38 2.0.0
- Python version: 3.6.8
- CUDA/cuDNN version: /
- GPU model and memory: /

**Describe the current behavior**

When using `model.fit` method with specified `validation_data`, the progress bar reports the validation (instead of training) dataset size as the total count in the progress bar. After the end of the epoch the values are correctly displayed.

The bug occurs when using tf.data datasets with unknown number of elements (for example when using generators).

Example output:

Epoch 1/10
1024/1024 [==============================] - 7s 7ms/step - loss: 23984083.2295 - accuracy: 0.1007 - val_loss: 0.0000e+00 - val_accuracy: 0.0000e+00
Epoch 2/10
1024/1024 [==============================] - 5s 5ms/step - loss: 25074923.1717 - accuracy: 0.0985 - val_loss: 30138559.6406 - val_accuracy: 0.1052
Epoch 3/10
 78/128 [=================>............] - ETA: 0s - loss: 28146315.7692 - accuracy: 0.1126

**Describe the expected behavior**

The progress bar should report the total number of elements infered from the training dataset, not from the validation dataset. The last line in the example output should be 78/1024, not 78/128.

**Code to reproduce the issue**

```python
import tensorflow as tf
from tensorflow.keras import layers
import numpy as np

# Model from tf Keras overview
model = tf.keras.Sequential([
    layers.Dense(64, activation='relu', input_shape=(32,)),
    layers.Dense(64, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model.compile(optimizer=tf.keras.optimizers.Adam(0.01),
              loss='categorical_crossentropy',
              metrics=['accuracy'])


# Data generator
def get_dataset(batch_size, num_batches):
    """ Get random tf.data dataset with given batch size and number of batches. """

    def _generator():
        for i in range(num_batches):
            data = np.random.random((batch_size, 32))
            labels = np.random.random((batch_size, 10))

            yield data, labels
            
    dataset = tf.data.Dataset.from_generator(
        _generator,
        (tf.float32, tf.float32),
        ((batch_size, 32), (batch_size, 10))
    )
    
    return dataset

# Create datasets
dataset = get_dataset(batch_size=32, num_batches=1024)
val_dataset = get_dataset(batch_size=32, num_batches=128)

# Train
model.fit(dataset, epochs=10,
          validation_data=val_dataset)
```

**Other info / logs**

The problem still occurs if `steps_per_epoch` is specified in the fit method. If `validation_steps` is provided, then the behaviour is correct. 

Seems like the computation of number of elements in the validation dataset somehow overrides the computed number from the training dataset.

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.0. Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/54fe6ec0fbb9a278f553acd31491b977/untitled395.ipynb). Is this the expected behavior?. Thanks!

**Comment by lojzezust** [comment]
Yeah this is the current behaviour, but not what should be happening. When you run the first example, you can see that first the dataset size is unknown (as it should be). After the first epoch is completed (1024 steps), evaluation is performed (128 steps). 

Then for other epochs the size of the evaluation dataset (128) is incorrectly used instead of the the training dataset (1024) as the total counter in the progress bar (it reads x/128). Only after the epoch si complete the progress bar corrects to 1024/1024. That's why after the training is done you can't really see anything wrong with it.

This is better shown in the second example, where you used 2 steps for training. Then it incorrectly shows 2/128 for every epoch instead of 2/2.

When you use `validation_step` like you did in the third example, then the behaviour is correct.

**Comment by btel** [comment]
I observe the same behaviour when using `tf.data.Dataset.from_generator`. However it seems to be fixed in the nighlty build. Try replacing the first line in the notebook with:

```!pip install tf-nightly```

**Comment by lojzezust** [comment]
Can confirm. Using the nightly build (2.1.0-dev20191122) I get the expected behaviour.

**Comment by ravikyram** [comment]
@lojzezust 
This issue is resolved with recent nightly builds. Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/71f7ae7f777f7a21b0b91a6c8a105aaa/untitled408.ipynb).Please, let us know we can close this issue?. Thanks!

**Comment by lojzezust** [comment]
Yeah, the issue is resoved in the nightly builds. Thanks. I will close the issue.

**Comment by muellesi** [comment]
Unfortunately it is not possible for me to use a nightly build. Can anyone tell me whether this issue is only a display problem or if the datasets really get mixed up during training? I can live with the wrong progress bar but the datasets getting mixed up would be a problem.

Exact package versions:

python=3.7.5
cudatoolkit=10.0.130=0
cudnn=7.6.4=cuda10.0_0
keras-applications=1.0.8=py_0
keras-preprocessing=1.1.0=py_1
tensorboard=2.0.0=pyhb38c66f_1
tensorflow=2.0.0=gpu_py37h57d29ca_0
tensorflow-base=2.0.0=gpu_py37h390e234_0
tensorflow-estimator=2.0.0=pyh2649769_0
tensorflow-gpu=2.0.0=h0d30ee6_0
tensorflow-probability=0.8.0=py_0
